Web Inspector: nodes that aren't visible should be visually distinguished in the DOM tree

dcrousso opened the pull request:
#### 7c836fc5a6c43742684b3f495cdc43f572b98ced
<pre>
Web Inspector: nodes that aren&apos;t visible should be visually distinguished in the DOM tree
<a href="https://bugs.webkit.org/show_bug.cgi?id=17219">https://bugs.webkit.org/show_bug.cgi?id=17219</a>
&lt;rdar://problem/5732820&gt;

Reviewed by Patrick Angle.

This will help developers more quickly identify whether a particular node in the Elements Tab is
something they care about (e.g. if they&apos;re adjusting layout, a `&lt;script&gt;` is probably irrelevant).

* Source/JavaScriptCore/inspector/protocol/CSS.json:
* Source/WebCore/inspector/InspectorInstrumentation.h:
(WebCore::InspectorInstrumentation::didChangeRendererForDOMNode): Renamed from `nodeLayoutContextChanged`.
* Source/WebCore/inspector/InspectorInstrumentation.cpp:
(WebCore::InspectorInstrumentation::didChangeRendererForDOMNodeImpl): Renamed from `nodeLayoutContextChangedImpl`.
* Source/WebCore/inspector/agents/InspectorCSSAgent.h:
* Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
(WebCore::InspectorCSSAgent::InspectorCSSAgent):
(WebCore::InspectorCSSAgent::reset):
(WebCore::layoutFlagContextTypeForRenderer): Renamed from `InspectorCSSAgent::layoutContextTypeForRenderer`.
(WebCore::InspectorCSSAgent::layoutFlagsForNode): Added.
(WebCore::pushChildrenNodesToFrontendIfLayoutFlagIsRelevant) Renamed from `pushChildrenNodesToFrontendIfLayoutContextTypePresent`.
(WebCore::InspectorCSSAgent::didChangeRendererForDOMNode): Renamed from `nodeLayoutContextTypeChanged`.
(WebCore::InspectorCSSAgent::nodesWithPendingLayoutFlagChangesTimerFired): Renamed from `layoutContextTypeChangedTimerFired`.
* Source/WebInspectorUI/UserInterface/Controllers/DOMManager.js:
(WI.DOMManager.prototype.nodeLayoutFlagsChanged): Renamed from `nodeLayoutContextTypeChanged`.
* Source/WebInspectorUI/UserInterface/Protocol/CSSObserver.js:
(WI.CSSObserver.prototype.nodeLayoutFlagsChanged): Renamed from `nodeLayoutContextTypeChanged`.
(WI.CSSObserver.prototype.nodeLayoutContextTypeChanged):
Expand `CSS.LayoutContextType` to be `CSS.LayoutFlag` (which is used in an array instead of as a
singular value), allowing for other layout information to be sent at the same time as layout context
type changes (currently the only other value is `Rendered`, which indicates whether the node has a
corresponding `WebCore::RenderObject` and is therefore rendered). Share the same instrumentation
point for layout context type changes and renderer changes, as both are closely related and can
therefore share the same timer (for coalescing rapid changes).
Drive-by: Move the logic for getting the `nodeId` into `nodesWithPendingLayoutFlagChangesTimerFired`
          so that we avoid extra work in this potentially hot code (i.e. before it was multiple
          branches/hashes and now it&apos;s just a single hash with all the other work deferred/coalesced).

* Source/WebCore/dom/Node.h:
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::notifyInspectorOfRendererChange): Added.
* Source/WebCore/rendering/RenderObject.h:
(WebCore::Node::setRenderer):
Add instrumentation inside `Node::setRenderer`. We can&apos;t use `InspectorInstrumentation` directly as
that causes linker problems, so instead we check `hasFrontends` to determine whether another
function is called, which then calls `InspectorInstrumentation` in the implementation.

* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::RenderFlexibleBox):
(WebCore::RenderFlexibleBox::~RenderFlexibleBox): Deleted.
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::RenderGrid):
(WebCore::RenderGrid::~RenderGrid): Deleted.
These instrumentation points are now handled by the above.

* Source/JavaScriptCore/inspector/protocol/DOM.json:
* Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
(WebCore::InspectorDOMAgent::buildObjectForNode):
* Source/WebInspectorUI/UserInterface/Models/DOMNode.js:
(WI.DOMNode):
(WI.DOMNode.prototype.get layoutFlags): Renamed from `get layoutContextType`.
(WI.DOMNode.prototype.set layoutFlags): Renamed from `set layoutContextType`.
(WI.DOMNode.prototype.get layoutContextType):
(WI.DOMNode.prototype.markDestroyed):
(WI.DOMNode.prototype.setAttributeValue):
(WI.DOMNode.prototype.showLayoutOverlay):
(WI.DOMNode.prototype.hideLayoutOverlay):
(WI.DOMNode.prototype._markUndoableState): Added.
(WI.DOMNode.prototype._makeUndoableCallback):
(WI.DOMNode.prototype._createLayoutOverlayColorSettingIfNeeded):
Replace the existing `layoutContextType` with the new `layoutFlags` when creating a `DOM.Node`.
Drive-by: Promisify `setAttributeValue` for use in tests.

* Source/WebInspectorUI/UserInterface/Views/DOMTreeElement.js:
(WI.DOMTreeElement.prototype.onattach):
(WI.DOMTreeElement.prototype.ondetach):
(WI.DOMTreeElement.prototype._updateLayoutBadge):
(WI.DOMTreeElement.prototype._handleLayoutFlagsChanged): Renamed from `_handleLayoutContextTypeChanged`.
* Source/WebInspectorUI/UserInterface/Views/DOMTreeOutline.css:
(.tree-outline.dom li:not(.rendered) &gt; span &gt; *:not(.html-comment)):
(.tree-outline.dom li:not(.rendered, .selected) &gt; span &gt; *:not(.html-comment), .tree-outline.dom:not(:focus-within) li.selected:not(.rendered) &gt; span &gt; *:not(.html-comment), body:is(.window-inactive, .window-docked-inactive) .tree-outline.dom li.selected:not(.rendered) &gt; span &gt; *:not(.html-comment)): Added.
Visually distinguish non-rendered nodes by making them greyscale.

* LayoutTests/inspector/css/nodeLayoutFlagsChanged-Rendered.html: Added.
* LayoutTests/inspector/css/nodeLayoutFlagsChanged-Rendered-expected.txt: Added.

* Source/WebInspectorUI/UserInterface/Views/LayoutDetailsSidebarPanel.js:
(WI.LayoutDetailsSidebarPanel.prototype.attached):
(WI.LayoutDetailsSidebarPanel.prototype.detached):
(WI.LayoutDetailsSidebarPanel.prototype._handleLayoutFlagsChanged): Renamed from `_handleLayoutContextTypeChanged`.
(WI.LayoutDetailsSidebarPanel.prototype._refreshNodeSets):
* LayoutTests/inspector/css/nodeLayoutContextTypeChanged.html:
* LayoutTests/inspector/css/setLayoutContextTypeChangedMode.html:
Use the renamed `WI.DOMNode.LayoutFlag` and `WI.DOMNode.Event.LayoutFlagsChanged`.

Canonical link: <a href="https://commits.webkit.org/252653@main">https://commits.webkit.org/252653@main</a>
</pre>

patrickangle reviewed (commented):
The whole concept of `nodeRendererChanged` is strikingly similar to `nodeLayoutContextTypeChanged`. Is there a reason we shouldn't have these two concepts be the same, and have the no-renderer case be another value in the (potentially renamed) `LayoutContextType` enum? Same would apply to the info on the node itself as well, `rendered` is really just an indication of a `layoutContextType` of a new value `none` (not to be confused with our current behavior of not putting a value in that field when it isn't a type we are interested in).

Also, can we have someone familiar with the render tree builder take a look - I've weary of any performance costs we are now incurring on every node while building the render tree (as opposed to previously only instrumenting grid/flexbox elements) and would certainly like someone's opinion on this that is more familiar with that area to make sure we get it right.

rniwa commented on Source/WebCore/dom/Node.h:
```diff
@@ -421,6 +421,7 @@ class Node : public EventTarget {
     // As renderer() includes a branch you should avoid calling it repeatedly in hot code paths.
+    // Make sure to call `InspectorInstrumentation::didChangeRendererForDOMNode` immediately after.
     RenderObject* renderer() const { return m_rendererWithStyleFlags.pointer(); }
     void setRenderer(RenderObject*); // Defined in RenderObject.h
```
Why doesn't setRenderer call InspectorInstrumentation::didChangeRendererForDOMNode instead?

dcrousso replied:
i'd tried this at first, but got a whole bunch of linker errors so i went with another approach just to see if i could get something working at all

i took another look today and think ive fixed all the issues :)

patrickangle commented on Source/JavaScriptCore/inspector/protocol/CSS.json:
```diff
@@ -479,12 +483,12 @@
-            "name": "nodeLayoutContextTypeChanged",
-            "description": "Called when a node's layout context type has changed.",
+            "name": "nodeLayoutFlagsChanged",
+            "description": "Called when the layout of a node changes.",
```
`Called when the layout of a node changes in a way that is important to Web Inspector.`?

patrickangle commented on Source/JavaScriptCore/inspector/protocol/CSS.json:
```diff
@@ -479,12 +483,12 @@
-                { "name": "nodeId", "$ref": "DOM.NodeId", "description": "Identifier of the node whose layout context type changed." },
-                { "name": "layoutContextType", "$ref": "LayoutContextType", "optional": true, "description": "The new layout context type of the node. When not provided, the <code>LayoutContextType</code> of the node is not a context for which Web Inspector has specific functionality." }
+                { "name": "nodeId", "$ref": "DOM.NodeId", "description": "Identifier of the node whose layout changed." },
+                { "name": "layoutFlags", "type": "array", "items": { "type": "string", "description": "CSS.LayoutFlag" }, "optional": true, "description": "Relevant information about the layout of the node. When not provided, the layout of the node is not important to Web Inspector." }
```
`"items": { "$ref": "LayoutFlag" }`?

dcrousso replied:
sadly we can't do that because the C++ protocol generator code doesn't know how to auto-convert a `JSON::ArrayOf` with an enum into the actual `JSON::ArrayOf<String>` that's sent to the frontend

and we can't create a specialization of `JSON::ArrayOf::addItem` because that exists in WTF, which is below `Inspector::Protocol::Helpers::getEnumConstantValue` (also that `getEnumConstantValue` is overloaded by WebDriver/`Automation` stuff, so we'd potentially have naming clashes)

this is kinda similar to how objects sent to the backend show up as `JSON::Object` instead of the type in the protocol, since we don't have sophisticated enough logic to auto-convert from the raw JSON to the `Inspector::Protocol` object

patrickangle commented on Source/JavaScriptCore/inspector/protocol/DOM.json:
```diff
@@ -67,7 +67,7 @@
                 { "name": "pseudoElements", "type": "array", "items": { "$ref": "Node" }, "optional": true, "description": "Pseudo elements associated with this node." },
                 { "name": "contentSecurityPolicyHash", "type": "string", "optional": true, "description": "Computed SHA-256 Content Security Policy hash source for given element." },
-                { "name": "layoutContextType", "$ref": "CSS.LayoutContextType", "optional": true, "description": "The layout context type of the node. When not provided, the <code>LayoutContextType</code> of the node is not a context for which Web Inspector has specific functionality." }
+                { "name": "layoutFlags", "type": "array", "items": { "type": "string", "description": "CSS.LayoutFlag" }, "optional": true, "description": "Relevant information about the layout of the node. When not provided, the layout of the node is not important to Web Inspector." }
```
`"items": { "$ref": "CSS.LayoutFlag" }`?

patrickangle commented on Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
```diff
@@ -935,26 +935,45 @@ Protocol::ErrorStringOr<void> InspectorCSSAgent::forcePseudoState(Protocol::DOM:
+
+    return layoutFlags;
+}
+
```
Style: extra newline

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/LayoutDetailsSidebarPanel.js:
```diff
@@ -171,17 +171,13 @@ WI.LayoutDetailsSidebarPanel = class LayoutDetailsSidebarPanel extends WI.DOMDet
-
-            default:
-                console.assert(false, "Unknown layout context type.", node, node.type);
-                break;
```
I guess this isn't reachable currently, but I'd suggest leaving it so if we add another valid layout context at some point we don't just silently do nothing here.

dcrousso replied:
oh lol oops i think this was removed in a previous design and i forgot to add it back

patrickangle reviewed (approved):
Awesome stuff - hopefully PLT et. al. come back neutral. So long as that's the case, the change looks good to me.

dcrousso committed:
Web Inspector: nodes that aren't visible should be visually distinguished in the DOM tree
https://bugs.webkit.org/show_bug.cgi?id=17219
<rdar://problem/5732820>

Reviewed by Patrick Angle.

This will help developers more quickly identify whether a particular node in the Elements Tab is
something they care about (e.g. if they're adjusting layout, a `<script>` is probably irrelevant).

* Source/JavaScriptCore/inspector/protocol/CSS.json:
* Source/WebCore/inspector/InspectorInstrumentation.h:
(WebCore::InspectorInstrumentation::didChangeRendererForDOMNode): Renamed from `nodeLayoutContextChanged`.
* Source/WebCore/inspector/InspectorInstrumentation.cpp:
(WebCore::InspectorInstrumentation::didChangeRendererForDOMNodeImpl): Renamed from `nodeLayoutContextChangedImpl`.
* Source/WebCore/inspector/agents/InspectorCSSAgent.h:
* Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
(WebCore::InspectorCSSAgent::InspectorCSSAgent):
(WebCore::InspectorCSSAgent::reset):
(WebCore::layoutFlagContextTypeForRenderer): Renamed from `InspectorCSSAgent::layoutContextTypeForRenderer`.
(WebCore::InspectorCSSAgent::layoutFlagsForNode): Added.
(WebCore::pushChildrenNodesToFrontendIfLayoutFlagIsRelevant) Renamed from `pushChildrenNodesToFrontendIfLayoutContextTypePresent`.
(WebCore::InspectorCSSAgent::didChangeRendererForDOMNode): Renamed from `nodeLayoutContextTypeChanged`.
(WebCore::InspectorCSSAgent::nodesWithPendingLayoutFlagChangesTimerFired): Renamed from `layoutContextTypeChangedTimerFired`.
* Source/WebInspectorUI/UserInterface/Controllers/DOMManager.js:
(WI.DOMManager.prototype.nodeLayoutFlagsChanged): Renamed from `nodeLayoutContextTypeChanged`.
* Source/WebInspectorUI/UserInterface/Protocol/CSSObserver.js:
(WI.CSSObserver.prototype.nodeLayoutFlagsChanged): Renamed from `nodeLayoutContextTypeChanged`.
(WI.CSSObserver.prototype.nodeLayoutContextTypeChanged):
Expand `CSS.LayoutContextType` to be `CSS.LayoutFlag` (which is used in an array instead of as a
singular value), allowing for other layout information to be sent at the same time as layout context
type changes (currently the only other value is `Rendered`, which indicates whether the node has a
corresponding `WebCore::RenderObject` and is therefore rendered). Share the same instrumentation
point for layout context type changes and renderer changes, as both are closely related and can
therefore share the same timer (for coalescing rapid changes).
Drive-by: Move the logic for getting the `nodeId` into `nodesWithPendingLayoutFlagChangesTimerFired`
          so that we avoid extra work in this potentially hot code (i.e. before it was multiple
          branches/hashes and now it's just a single hash with all the other work deferred/coalesced).

* Source/WebCore/dom/Node.h:
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::notifyInspectorOfRendererChange): Added.
* Source/WebCore/rendering/RenderObject.h:
(WebCore::Node::setRenderer):
Add instrumentation inside `Node::setRenderer`. We can't use `InspectorInstrumentation` directly as
that causes linker problems, so instead we check `hasFrontends` to determine whether another
function is called, which then calls `InspectorInstrumentation` in the implementation.

* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::RenderFlexibleBox):
(WebCore::RenderFlexibleBox::~RenderFlexibleBox): Deleted.
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::RenderGrid):
(WebCore::RenderGrid::~RenderGrid): Deleted.
These instrumentation points are now handled by the above.

* Source/JavaScriptCore/inspector/protocol/DOM.json:
* Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
(WebCore::InspectorDOMAgent::buildObjectForNode):
* Source/WebInspectorUI/UserInterface/Models/DOMNode.js:
(WI.DOMNode):
(WI.DOMNode.prototype.get layoutFlags): Renamed from `get layoutContextType`.
(WI.DOMNode.prototype.set layoutFlags): Renamed from `set layoutContextType`.
(WI.DOMNode.prototype.get layoutContextType):
(WI.DOMNode.prototype.markDestroyed):
(WI.DOMNode.prototype.setAttributeValue):
(WI.DOMNode.prototype.showLayoutOverlay):
(WI.DOMNode.prototype.hideLayoutOverlay):
(WI.DOMNode.prototype._markUndoableState): Added.
(WI.DOMNode.prototype._makeUndoableCallback):
(WI.DOMNode.prototype._createLayoutOverlayColorSettingIfNeeded):
Replace the existing `layoutContextType` with the new `layoutFlags` when creating a `DOM.Node`.
Drive-by: Promisify `setAttributeValue` for use in tests.

* Source/WebInspectorUI/UserInterface/Views/DOMTreeElement.js:
(WI.DOMTreeElement.prototype.onattach):
(WI.DOMTreeElement.prototype.ondetach):
(WI.DOMTreeElement.prototype._updateLayoutBadge):
(WI.DOMTreeElement.prototype._handleLayoutFlagsChanged): Renamed from `_handleLayoutContextTypeChanged`.
* Source/WebInspectorUI/UserInterface/Views/DOMTreeOutline.css:
(.tree-outline.dom li:not(.rendered) > span > *:not(.html-comment)):
(.tree-outline.dom li:not(.rendered, .selected) > span > *:not(.html-comment), .tree-outline.dom:not(:focus-within) li.selected:not(.rendered) > span > *:not(.html-comment), body:is(.window-inactive, .window-docked-inactive) .tree-outline.dom li.selected:not(.rendered) > span > *:not(.html-comment)): Added.
Visually distinguish non-rendered nodes by making them greyscale.

* LayoutTests/inspector/css/nodeLayoutFlagsChanged-Rendered.html: Added.
* LayoutTests/inspector/css/nodeLayoutFlagsChanged-Rendered-expected.txt: Added.

* Source/WebInspectorUI/UserInterface/Views/LayoutDetailsSidebarPanel.js:
(WI.LayoutDetailsSidebarPanel.prototype.attached):
(WI.LayoutDetailsSidebarPanel.prototype.detached):
(WI.LayoutDetailsSidebarPanel.prototype._handleLayoutFlagsChanged): Renamed from `_handleLayoutContextTypeChanged`.
(WI.LayoutDetailsSidebarPanel.prototype._refreshNodeSets):
* LayoutTests/inspector/css/nodeLayoutContextTypeChanged.html:
* LayoutTests/inspector/css/setLayoutContextTypeChangedMode.html:
Use the renamed `WI.DOMNode.LayoutFlag` and `WI.DOMNode.Event.LayoutFlagsChanged`.

Canonical link: https://commits.webkit.org/252653@main
